### ISSUE TITLE
Better C++ documentation

### ISSUE THREAD

**Issue by noah1510** [libvips/libvips#1781]
At the moment the C++ documentation is terrible.
It more or less only helps if you already know the API and I think for people new to programming (at least new to C/C++ programming) have no idea how they could use the library for their program.

Basically the whole documentation should be written in a way that somebody who knows almost nothing about C programming can write a C++ program that utilizes libvips.

The following list are things I think are missing at the moment:

* [ ] a description of the examples and what needs to be passed to execute the program
* [ ] a list of all members and methods of the VImage class and their use (doxygen output would be enough)
* [ ] a list of things in the vips namespace and their use

These things should be part of the official documentation.

If you feel like there is anything missing I will add it to the list.

**Comment by jcupitt** [comment]
Hello @noah1510,

Yes, I can see doxygen output would help. It ought to be integrated into the build system.

Once that's in, extra doc comments could be added to eg. VImage8.h.

We could add stuff to the Python that generates the C++ API to also generate extra doxy comments.

I agree that the "libvips from C++" chapter could be improved quite a bit.

**Comment by jaubourg** [comment]
Just sharing my experience here.

Using the C++ API has been a struggle. Always having to reference back to the C API is very confusing, especially regarding life cycles, memory management and, of course, finding methods.

If anything, a C++ doc that would not reference the C API would be a plus, though it's obviously a lot more work.

As a side note, searching on google for libvips C++ info is close to impossible so SEO could probably use some help.

So, I'm getting there but it's been hard. That being said, I'm using other C++ APIs for my service and having an opaque or even non existent documentation seems to be pretty common in the C++ ecosystem. Makes for quite a contrast when coming back from the JavaScript ecosystem.

Anyway, your responsiveness and willingness to engage in issues is a big plus in my book (especially when you're asked questions that end up in a "oh snap! Of course" moment from the one asking). So thank you for the lib and your dedication!

**Comment by jcupitt** [comment]
I made a branch for experimenting with doxygen:

https://github.com/libvips/libvips/tree/add-doxygen

If anyone knows doxy well and would like to help configure it, that would be great!

**Comment by noah1510** [comment]
I just created a doxyfile to gernerate the documentation, but because the methods are not documented, there is not much to see.
For doxygen to be really useful, everything should be commented in javadoc style which will be parsed by doxygen.
I think it will take a lot of work to get a good amount of coverage.

There is an extention for vs code to [automatically generate documentation](https://open-vsx.org/extension/cschlosser/doxdocgen) but the description still have to be manually written.

**Comment by jcupitt** [comment]
There are doc comments in the headers generated from libvips introspection. I ran doxy to make this:

http://www.rollthepotato.net/~john/html/classVImage.html

See eg.:

http://www.rollthepotato.net/~john/html/classVImage.html#adf824d6bd39159b8f67ab47b710181c1

We could generate docs for the options too, that would be useful.

There's also XML with all the C docs as well, we could perhaps paste that in somehow.

**Comment by jcupitt** [comment]
We have this XML for the `embed` operator:

<details> 
<summary>xml for embed</summary>

```xml
      <method name="embed" c:identifier="vips_embed" introspectable="0">
        <doc xml:space="preserve"
             filename="conversion/embed.c"
             line="678">Optional arguments:

* @extend: #VipsExtend to generate the edge pixels (default: black)
* @background: #VipsArrayDouble colour for edge pixels

The opposite of vips_extract_area(): embed @in within an image of size
@width by @height at position @x, @y.

@extend
controls what appears in the new pels, see #VipsExtend.

See also: vips_extract_area(), vips_insert().</doc>
        <source-position filename="include/vips/conversion.h" line="150"/>
        <return-value transfer-ownership="none">
          <doc xml:space="preserve"
               filename="conversion/embed.c"
               line="701">0 on success, -1 on error.</doc>
          <type name="gint" c:type="int"/>
        </return-value>
        <parameters>
          <instance-parameter name="in" transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="680">input image</doc>
            <type name="Image" c:type="VipsImage*"/>
          </instance-parameter>
          <parameter name="out"
                     direction="out"
                     caller-allocates="0"
                     transfer-ownership="full">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="681">output image</doc>
            <type name="Image" c:type="VipsImage**"/>
          </parameter>
          <parameter name="x" transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="682">place @in at this x position in @out</doc>
            <type name="gint" c:type="int"/>
          </parameter>
          <parameter name="y" transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="683">place @in at this y position in @out</doc>
            <type name="gint" c:type="int"/>
          </parameter>
          <parameter name="width" transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="684">@out should be this many pixels across</doc>
            <type name="gint" c:type="int"/>
          </parameter>
          <parameter name="height" transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="685">@out should be this many pixels down</doc>
            <type name="gint" c:type="int"/>
          </parameter>
          <parameter name="..." transfer-ownership="none">
            <doc xml:space="preserve"
                 filename="conversion/embed.c"
                 line="686">%NULL-terminated list of optional named arguments</doc>
            <varargs/>
          </parameter>
        </parameters>
      </method>
```

</details>

And we can get this from introspection:

<details>
<summary>introspection for embed</summary>

```
$ vips embed
embed an image in a larger image
usage:
   embed in out x y width height [--option-name option-value ...]
where:
   in           - Input image, input VipsImage
   out          - Output image, output VipsImage
   x            - Left edge of input in output, input gint
			default: 0
			min: -1000000000, max: 1000000000
   y            - Top edge of input in output, input gint
			default: 0
			min: -1000000000, max: 1000000000
   width        - Image width in pixels, input gint
			default: 1
			min: 1, max: 1000000000
   height       - Image height in pixels, input gint
			default: 1
			min: 1, max: 1000000000
optional arguments:
   extend       - How to generate the extra pixels, input VipsExtend
			default: black
			allowed: black, copy, repeat, mirror, white, background
   background   - Color for background pixels, input VipsArrayDouble
```

</details>

We'll need to write doc comments for things like `VImage::new_from_file`.

There's the `libvips from C++` chapter in the docs we could paste in too. Does doxygen allow markdown in comments?

**Comment by noah1510** [comment]
> Does doxygen allow markdown in comments?

Yes you can use markdown in doxygen, you can also reference other objects

**Comment by noah1510** [comment]
I just found out that most of the documentation was missing because I did not run ./configure before building the documentation, now the documentation is actually useful.

**Comment by jcupitt** [comment]
Build systems are all terrible :(

**Comment by jcupitt** [comment]
I added something to generate docs for the optional parameters too, eg.:

http://www.rollthepotato.net/~john/html/classVImage.html#adf824d6bd39159b8f67ab47b710181c1

And I've integrated doxygen into the build system, so it's run on "make" and the generated html is installed to `$prefix/share/doc/vips/html` on "make install".

To do:

- [ ] We should write an intro page to the docs and paste it in as a comment somewhere. The "libvips from C++" chapter in the main docs would be a good starting point.
- [ ] Someone needs to sit down and write doc comments for the hand-written parts of the C++ binding. There are about 250 of these, according to doxygen. Most are trivial.
- [ ] The thing that makes the API docs on the libvips website should include these formatted C++ docs.

I'm out of time for now, so any PRs would be very welcome.
